pkg/controller/incompatible_operator_controller.go

```diff
@@ -0,0 +1,168 @@
+		nextOCPMinorVersion:    nextOCPMinorVersion,
+		kubeclient:             kubeclient,
+		clusterExtensionClient: clusterExtensionClient,
+		operatorClient:         operatorClient,
+		logger:                 klog.NewKlogr().WithName(name),
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(operatorClient.Informer(), clusterExtensionClient.Informer().Informer()).ToController(name, eventRecorder)
+}
+
+func (c *incompatibleOperatorController) sync(ctx context.Context, _ factory.SyncContext) error {
+	c.logger.Info("sync started")
+	defer c.logger.Info("sync finished")
+
+	incompatibleOperators, err := c.getIncompatibleOperators()
+	if err != nil {
+		c.logger.V(4).Error(err, "error checking incompatible operators")
+		return err
+	}
+
+	var updateStatusFn v1helpers.UpdateStatusFunc
+	if len(incompatibleOperators) > 0 {
+		updateStatusFn = v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "InstalledOLMOperatorsUpgradeable",
+			Status:  operatorv1.ConditionFalse,
+			Reason:  reasonIncompatibleOperatorsInstalled,
+			Message: strings.Join(incompatibleOperators, ","),
+		})
+	} else {
+		updateStatusFn = v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "InstalledOLMOperatorsUpgradeable",
+			Status: operatorv1.ConditionTrue,
+		})
+	}
+
+	if _, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, updateStatusFn); updateErr != nil {
+		c.logger.V(4).Info(fmt.Sprintf("Error updating operator condition status: %v", updateErr))
+		return updateErr
+	}
+	return nil
+}
+
+func (c *incompatibleOperatorController) getIncompatibleOperators() ([]string, error) {
+	var incompatibleOperators []string
+
+	ceList, err := c.clusterExtensionClient.Informer().Lister().List(labels.NewSelector())
+	if err != nil {
+		c.logger.V(4).Error(err, "Error listing cluster extensions")
+		return nil, err
+	}
+
+	store := c.buildHelmStore(c.kubeclient.CoreV1().Secrets("openshift-operator-controller"))
+
+	// Get all ClusterExtensions incompatible with next Y-stream
+	for _, obj := range ceList {
+		ce, ok := obj.(*unstructured.Unstructured)
+		if !ok {
+			c.logger.V(4).Error(nil, fmt.Sprintf("Unable to Cast Object to unstructured.Unstructured. Object is of type %T", obj))
+			return nil, nil
+		}
+
+		rel, err := store.Deployed(ce.GetName())
+		if errors.Is(err, driver.ErrNoDeployedReleases) {
+			c.logger.Info("Cluster Extension not yet deployed - will check again later")
+			continue
+		}
+		if err != nil {
+			c.logger.V(4).Error(err, "Error returning the last deployed release")
+			return nil, err
+		}
+
+		if rel.Chart == nil || rel.Chart.Metadata == nil {
+			continue
+		}
+		props, err := propertyListFromPropertiesAnnotation(rel.Chart.Metadata.Annotations["olm.properties"])
+		if err != nil {
+			c.logger.V(4).Error(err, "Error converting olm.properties")
+			return nil, err
+		}
+		for _, p := range props {
+			if p.Type == maxOpenShiftVersionProperty {
```

Comment:
We probably need to count the number of times we see this property and emit an error if we see it more than once.

Reply:
(for a single bundle)